libs/langchain-google-cloud-sql-pg/src/engine.ts

```diff
@@ -0,0 +1,315 @@
+
+  pool: knex.Knex;
+
+  static connector: Connector;
+
+  constructor(key: symbol, pool: knex.Knex) {
+    if (key !== PostgresEngine._createKey) {
+      throw Error("Only create class through 'create' method!");
+    }
+    this.pool = pool;
+  }
+
+  /**
+   * @param projectId Required - GCP Project ID
+   * @param region Required - Postgres Instance Region
+   * @param instance Required - Postgres Instance name
+   * @param database Required - Database name
+   * @param ipType Optional - IP address type. Defaults to IPAddressType.PUBLIC
+   * @param user Optional - Postgres user name. Defaults to undefined
+   * @param password Optional - Postgres user password. Defaults to undefined
+   * @param iamAccountEmail Optional - IAM service account email. Defaults to undefined
+   * @returns PostgresEngine instance
+   */
+
+  static async fromInstance(
+    projectId: string,
+    region: string,
+    instance: string,
+    database: string,
+    {
+      ipType = IpAddressTypes.PUBLIC,
+      user,
+      password,
+      iamAccountEmail,
+    }: PostgresEngineArgs = {}
+  ): Promise<PostgresEngine> {
+    let dbUser: string;
+    let enableIAMAuth: boolean;
+
+    if ((!user && password) || (user && !password)) {
+      // XOR for strings
+      throw Error(
+        "Only one of 'user' or 'password' were specified. Either " +
+          "both should be specified to use basic user/password " +
+          "authentication or neither for IAM DB authentication."
+      );
+    }
+
+    // User and password are given so we use the basic auth
+    if (user !== undefined && password !== undefined) {
+      enableIAMAuth = false;
+      dbUser = user!;
+    } else {
+      enableIAMAuth = true;
+      if (iamAccountEmail !== undefined) {
+        dbUser = iamAccountEmail;
+      } else {
+        // Get application default credentials
+        const auth = new GoogleAuth({
+          scopes: "https://www.googleapis.com/auth/cloud-platform",
+        });
+        // dbUser should be the iam principal email by passing the credentials obtained
+        dbUser = await getIAMPrincipalEmail(auth);
+      }
+    }
+
+    PostgresEngine.connector = new Connector({ userAgent: USER_AGENT });
+    const clientOpts = await PostgresEngine.connector.getOptions({
+      instanceConnectionName: `${projectId}:${region}:${instance}`,
+      ipType,
+      authType: enableIAMAuth ? AuthTypes.IAM : AuthTypes.PASSWORD,
+    });
+
+    const dbConfig: knex.Knex.Config = {
+      client: "pg",
+      connection: {
+        ...clientOpts,
+        ...(password ? { password } : {}),
+        user: dbUser,
+        database,
+      },
+    };
+
+    const engine = knex(dbConfig);
+
+    return new PostgresEngine(PostgresEngine._createKey, engine);
+  }
+
+  /**
+   * Create a PostgresEngine instance from an Knex instance.
+   *
+   * @param engine knex instance
+   * @returns PostgresEngine instance from a knex instance
+   */
+  static async fromEngine(engine: knex.Knex) {
```

Comment:
nit: `PostgresEngine.fromEngine` is a little strange

Reply:
This adds a lot of flexibility for developers to use a connection pool they are already using and have configured for their application

Reply:
Ya just the naming is funny

Reply:
You are right, this didn't translate well from python. I have updated it to be `fromPool`